### PR TITLE
Stagnation Pressure Feature: dynamic pressure q_inf as input channel

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Add the freestream dynamic pressure `q_inf = 0.5 * Umag²` as a single new input channel broadcast to all nodes. This provides a **Bernoulli-derived pressure baseline** that tells the model what pressure deviation to EXPECT at each sample's velocity regime.

**Why this should work:** The model already has Umag as a global scalar, but q_inf = 0.5 * Umag² is a nonlinear derived quantity that the first linear layer must learn to compute. Providing it explicitly removes this burden. For the residual prediction framework (already merged), having a physics-informed baseline improves what the model learns to correct. This is different from the Bernoulli consistency LOSS (#2224, failed) — that imposed a hard constraint. This is just a feature, with no loss constraint.

**Category:** Input feature addition — the only category with consistent wins (wake deficit -4.1%, TE coord frame -5.4%).

## Instructions

Add 1 new input channel. Very simple implementation.

### Implementation

1. **Find where Umag is available** in the data pipeline. Read `prepare_multi.py` to locate Umag in the input tensor or global features.

2. **Compute and append q_inf:**
```python
# Umag is already a per-sample scalar in the input features
umag = x[:, :, umag_channel]  # [B, N] — broadcast Umag at all nodes
q_inf = 0.5 * umag ** 2      # [B, N] — dynamic pressure
q_inf_feat = q_inf.unsqueeze(-1)  # [B, N, 1]

x = torch.cat([x, q_inf_feat], dim=-1)  # [B, N, 25]
```

3. **Add flag** `--stagnation_feature`. Update `input_dim` to add 1 when set.

4. **Note:** Since the baseline already normalizes pressure by q_inf via `_phys_norm()` for the targets, providing q_inf as an INPUT feature gives the model explicit access to the normalization denominator. This should help especially for the residual prediction path.

### Training command (seed 42)
```bash
cd cfd_tandemfoil && python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/stagnation-pressure-s42" \
  --wandb_group "stagnation-pressure-feature" \
  --stagnation_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

Seed 73: same with `--seed 73` and adjusted `--wandb_name`.

## Baseline (PR #2251, 2-seed avg)

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.891** | < 11.89 |
| **p_oodc** | **7.561** | < 7.56 |
| **p_tan** | **28.118** | < 28.12 |
| p_re | 6.364 | < 6.36 |

W&B baseline: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Seed 42 (W&B: 7iz07fx6)
| Metric | Value | Baseline | Delta |
|--------|-------|----------|-------|
| **p_in** | **23.9** | 11.891 | +101.2% |
| **p_oodc** | **18.2** | 7.561 | +140.7% |
| **p_tan** | **42.5** | 28.118 | +51.2% |
| **p_re** | **12.6** | 6.364 | +98.0% |
| val_loss | 0.674 | — | — |
| best_epoch | 148 | — | — |
| peak VRAM | 46.8 GB | — | — |

### Seed 73 (W&B: py14l7mk)
| Metric | Value | Baseline | Delta |
|--------|-------|----------|-------|
| **p_in** | **50.8** | 11.891 | +327.3% |
| **p_oodc** | **48.8** | 7.561 | +545.4% |
| **p_tan** | **69.7** | 28.118 | +147.9% |
| **p_re** | **33.8** | 6.364 | +431.2% |
| val_loss | 2.136 | — | — |
| best_epoch | 78 | — | — |
| peak VRAM | 45.8 GB | — | — |

### 2-Seed Average
| Metric | 2-seed avg | Baseline | Delta |
|--------|-----------|----------|-------|
| **p_in** | **37.35** | 11.891 | +214.2% |
| **p_oodc** | **33.50** | 7.561 | +343.1% |
| **p_tan** | **56.10** | 28.118 | +99.5% |
| **p_re** | **23.20** | 6.364 | +264.6% |

### Training Commands
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/stagnation-pressure-s42" \
  --wandb_group "stagnation-pressure-feature" \
  --stagnation_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73: same with --seed 73 and --wandb_name "frieren/stagnation-pressure-s73"
```

### What happened — NEGATIVE RESULT

**The stagnation pressure feature (q_inf = 0.5 * Umag²) significantly degraded performance.** All metrics are 2-5x worse than baseline.

**Root cause analysis:**

1. **Un-normalized large-magnitude feature:** The raw q_inf = 0.5 * Umag² was appended directly to x without normalization. For the CFD domain, Umag ranges from ~1 to ~100+ m/s, so q_inf ranges from ~0.5 to ~5000+. This creates a feature with vastly different scale from the other inputs (which are standardized). The preprocess MLP has to learn to handle this extreme dynamic range.

2. **Training instability (s73):** s73 suffered a catastrophic loss spike at epoch 80 (val_tandem jumped to 1145) and never fully recovered. The best model was from epoch 78 (before the spike), with very poor surface MAE. This spike is likely caused by the un-normalized q_inf creating occasional gradient explosions when high-Re samples produce large q values.

3. **s42 was more stable** — no catastrophic spikes, excellent convergence trajectory, but still much worse than baseline on surface MAE. This suggests the feature itself is harmful, not just the instability.

4. **Verification NaN issue:** Both seeds showed NaN in the manual verification step. This doesn't affect training metrics but indicates a numerical issue in the verification path related to q_inf computation.

5. **Why the hypothesis failed:** The PR hypothesized that providing q_inf explicitly would save the model from learning the nonlinear transformation 0.5*Umag². However, the model already has log_Re as input and the existing _phys_norm already handles pressure normalization by q. Adding raw q_inf as an input feature is actually counterproductive — it introduces a high-variance, un-normalized feature that disrupts the learned representations. Unlike TE coord frame and wake deficit features (which are geometric and bounded), q_inf varies by orders of magnitude across samples.

### Suggested follow-ups

1. **If retried:** Normalize q_inf before appending (e.g., log(q) or standardize using training stats). The raw value is the likely killer.
2. **Alternative:** Instead of q_inf as input, try a Cp-prediction head that receives q_inf as a denormalization factor — i.e., predict Cp = p/q directly, which is bounded and well-conditioned.
3. **More promising directions:** Focus on architectural improvements (attention patterns, loss functions) rather than input features derived from targets, which risk information leakage and scale issues.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2291 because ADVISOR: Closing — CATASTROPHIC regression: p_in +218%, p_oodc +338%, p_tan +101%, p_re +261%. Root cause: un-normalized q_inf = 0.5*Umag² with extreme dynamic range (0.5 to 5000+) appended to standardized features. This caused gradient explosions (s73 had a loss spike at epoch 80, never recovered). The feature is also redundant — Umag is already in the input features, and q_inf = 0.5*Umag² is a trivial transformation the model can learn. If revisiting, would need log-scaling or standardization, but the redundancy with existing Umag makes it pointless. W&B: 7iz07fx6 (s42), py14l7mk (s73)..
